doc/decisions/0_drafts/buffer_struct.md

```diff
@@ -0,0 +1,189 @@
+# Buffer Struct
+
+## Problem
+
+For this decision a buffer is a fixed-length sequence of arbitrary bytes.
+In other words, it is a pointer to data combined with the size of this data.
+
+Many parts of Elektra use such buffers, but there is currently no uniform API for all of these.
```

Comment:
This by itself is not a problem. We should not provide an API for everything (e.g. strings, float serialization, etc.) but only for our use cases #4686.

But I agree that this area has problems and improvements are possible. Please specify what you aim for:

- optimizations? the size could be part of the allocated buffer, then it wouldn't need space if the buffer is NULL.
  Also the size field could grow as needed (with a bit indicating if further bytes also hold the size).
- less bugs? buffer resizing etc. is error-prone and we basically have no tests in this area

Also the urgency should be described. I assume it would be good to have for COW? @atmaxinger what are your thoughts on that?

Reply:
I know this is not really a problem (see https://github.com/ElektraInitiative/libelektra/pull/4683#discussion_r1023212527). I would have just made a proposal issue, but I wanted to avoid the inevitable "make a decision PR, please".

I have to ask, have you read the rest of the file, or are you waiting for a "clear problem"? If you haven't read the file, I would really encourage you to do so. There isn't really a clear problem here, but I think having the `ElektraBuffer` struct would make many things better.

I think when you read the whole document and see how much effort is put into the different options, it also becomes much more obvious that this is just a proposal.

> atmaxinger what are your thoughts on that?

He already left a review, see https://github.com/ElektraInitiative/libelektra/pull/4683#pullrequestreview-1181540342

Reply:
> the size could be part of the allocated buffer

Honestly I think this adds a lot of complexity for a very minor gain. 

> buffer resizing etc. is error-prone 

I don't see how this would make it any better. It just groups the size of the buffer with the buffer itself. Maybe we could provide reusable methods for buffer resizing for this, then the logic only has to be implemented and tested once. But that would require that all usecases for buffer resizing are the same.

> I assume it would be good to have for COW

I'm not sure whether we gain much. Functionality wise, it has not impact whatsoever. The only way it would matter if theses buffer structs would also have reference counting, then we could just use those. But as @kodebach mentioned, reference counting is orthogonal to this. So I don't see a big correlation between this an COW.

Reply:
I tried to ignore the ideas for "what you aim for", because there is no goal, but some further comments:

1. Size as part of buffer: _Always_ doing this, is absolutely **NOT** a goal. It completely breaks many of the use cases I had in mind with the public struct.
    However, there are options, e.g.:
    ```c
    ElektraBuffer * elektraBufferAlloc(size_t size) {
        if (size == 0) return NULL;
        ElektraBuffer * buffer = malloc (sizeof (ElektraBuffer) + size);
        buffer->size = size;
        buffer->data = buffer + 1; // pointer arithmetic in terms of sizeof(ElektraBuffer)
        return buffer;
    }
    ```
    I _intentionally_ did not mention this in the file, because buffers created like this are a special use case of `ElektraBuffer` and providing such an API would just lead to confusion. However, for COW -- because it would all be private API -- we could do _roughly_:

    ```c
    typedef struct {
       ElektraBuffer buffer;
       uint16_t refcount;
    } RcBuffer;

    static RcBuffer * rcBufferAlloc(size_t size) {
        if (size == 0) return NULL;
        RcBuffer * rc = elektraMalloc (sizeof (RcBuffer) + size);
        rc->buffer.size = size;
        rc->buffer.data = rc + 1; // pointer arithmetic in terms of sizeof(RcBuffer)
        rc->refcount = 1;
        return rc;
    }
    
    void keySetName (Key * key, void * name, size_t size) {
        rcBufferRelease(key->name);
        key->name = allocRcBuffer(size);
        memcpy (key->name->buffer.data, name, size);
    }
    ```
2. Buffer resizing: Again not intentionally omitted. Private allocation/resizing APIs using `ElektraBuffer` are of course possible. But for a public API this could be confusing, because with the public struct this is possible:
   ```c
    char data[] = "some data";
    ElektraBuffer buffer = {.data = data, .size = sizeof(data)};
   ```
   In such a case using a resizing API `resizeBuffer(&buffer)` would obviously fail. The only way I can see having such things as public API would be with e.g.
   ```c
   typedef ElektraBuffer ElektraHeapBuffer;
   ```

3. COW: As stated, there are some ways of making the internal private APIs simpler and maybe removing some duplication. But I agree that this can definitely be done after COW is done and we shouldn't delay COW because of this decision.

Reply:
> This by itself is not a problem. We should not provide an API for everything

I've just read our comment again. Maybe this will help, if it does, please confirm:

```suggestion
Many parts of Elektra use such buffers, but there is currently no uniform way of using them throughout the codebase.
```

To clarify further (if I have to, I'll add something like this to the file too):
- With API I **did not** mean public API. If I did I would have written "public API".
- Yes, I do suggest possible uses in the public API and I believe using these structs in the public API would be good. But that is not what this decision is about. Stating a problem for that would have been much simpler. This decision is just about the introduction of a uniform way of handling buffers, because currently we handle buffers in whatever way the author of a particular piece of code felt like.
- This is not just about string-like buffers. It's more about the fact that C doesn't remember the size of an array. The decision could also be applied to the `Key *` array in a `KeySet`. Whether or not that would be a good idea, is a different question. But there are enough string-like buffers that the decision still makes sense without it.